### PR TITLE
Enhancements for Carrier Billing v0.2 - Part II

### LINKED ISSUE #109: Default values for page and perPage query params in `retrievePayments` endpoint
**Problem description**
Based on the feedback of our developers about API implementation, they have feedback that the abscense of default values for page and perPage query params in `retrievePayments` endpoint ( GET /payments) can lead to different implementations when no other query parameter (i.e. filter criteria) is indicated. 

**Possible evolution**

As per their feedback in order to have an aligment regarding this scenario, they comment it can be settled for default values:
page: 1
perPage: 10

Then, when no explicit indication and no other query parameters indicated, API can return the 10 latest payments which is a good UX approach and also can have a good perfomance (API response time)

Another consideration is to set a maximum value for perPage (they have commneted to have a limit of 100, in order to avoid long response times)

**Additional context**

Maybe these considerations could be also commented in Commonalities to set-up this reference values across any API that can have pagination in future

### LINKED ISSUE #110: Aligment in "transactionOperationStatus" transitions and potential need of new Exception PAYMENT_DENIED
**Problem description**

It has been detected that in 2-STEP Payment there may be conditions where a given payment can be denied, e.g. User phone_number is blocked for any reason in the OB.

To deal with scenarios an exception is needed. The proposal is to be HTTP 403 CARRIER_BILLING.PAYMENT_DENIED

Related to this, think is important we can have alignment in the understanding of "transactionOperationStatus" so maybe other impacts are required and take advantage of this to reflect appropiate documentation in the context of [Issue#106](https://github.com/camaraproject/CarrierBillingCheckOut/issues/106)

Then:

**1-STEP PAYMENT**
- SYNC Process: `paymentId` is created, `transactionOperationStatus` is `succeeded`. It may also makes sense here PAYMENT_DENIED (even paymentId is not created)?

- ASYNC Process: `paymentId` is created, `transactionOperationStatus` is `processing`. Next steps would be:
> If Payment OK, then `transactionOperationStatus` is `succeeded`
> If Payment NOK, then `transactionOperationStatus` is `denied`. Therefore I think new event type "PAYMENT_DENIED" would be needed (it would also apply for 2-STEP, but see also relevant in 1-STEP)

**2-STEP PAYMENT**
- SYNC Process:
> `preparePayment`:  `paymentId` is created, `transactionOperationStatus` is `reserved`. It may also makes sense here PAYMENT_DENIED (even paymentId is not created)?
> `confirmPayment`: At this point, If Payment OK, then `transactionOperationStatus` is `succeeded` or If Payment NOK, then `transactionOperationStatus` is `denied`. 
> `cancelPayment`:  `paymentId` is created, `transactionOperationStatus` is `cancelled`

- ASYNC Process:
> `preparePayment`:  `paymentId` is created, `transactionOperationStatus` is `processing`. It may also makes sense here PAYMENT_DENIED (even paymentId is not created)?
> `confirmPayment`: At this point, 202 Accepted and If Payment OK, then `transactionOperationStatus` is `succeeded` (indicated by event) or If Payment NOK, then `transactionOperationStatus` is `denied` (indicated by event)  
> `cancelPayment`:  `paymentId` is created, `transactionOperationStatus` is `cancelled`

cc @bigludo7 

**Possible evolution**

Once aligned behaviour.
1) Document this "transactionOperationStatus" information associated to the track of Improving Documentation [Issue#106](https://github.com/camaraproject/CarrierBillingCheckOut/issues/106)
2) Consider new event type "PAYMENT_DENIED" (aligned with Commonalities Notifications Design)
3) Add new exception 403 "CARRIER_BILLING.PAYMENT_DENIED" in several methods:
- `confirmPayment`
- Based on aligment it may also apply to `createPayment` and `preparePayment` 

**Additional context**
N/A

### LINKED ISSUE #111: Enhancements in GET payment endpoints
**Problem description**

Be more accurate in the description of these endpoints to indicate they are refering in the context of a user phone number.
Add an exception when phone number cannot be deducted from Access Token context

**Possible evolution**

**1) Endpoint GET /payments**

_Update endpoint description:_

Retrieve a list of payments **of a given phone number** and their details based on some filtering criteria. Only payments recently performed (e.g. in the last hour) by the API client invoking this method will be returned; time window for availability will be a configuration in API Server.

- The operation is executed for the user whose sub is in the access token used to call this endpoint, and for the phone number also deducted from the information included in the access token.
- In case the user phone number cannot be obtained from access token information `HTTP 403` exception will be returned

_Add this exception to this endpoint:_
- 403
- "CARRIER_BILLING.INVALID_TOKEN_CONTEXT"
- "Phone Number cannot be deducted from Authorization context"

**2) Endpoint GET /payments/{paymentId}**

_Update endpoint description:_
Retrieve payment details for a given payment implicitly associated to a given phone number

- The operation is executed for the user whose sub is in the access token used to call this endpoint, and for the phone number also deducted from the information included in the access token.
- In case the user phone number cannot be obtained from access token information `HTTP 403` exception will be returned

_Add this exception to this endpoint:_
- 403
- "CARRIER_BILLING.INVALID_TOKEN_CONTEXT"
- "Phone Number cannot be deducted from Authorization context"

**Additional context**
N/A

### LINKED ISSUE #116: [Optional] validate Payment procedure in 2-STEP
**Problem description**
There can be the business scenario of 2-STEP payment where a specific payment needs to be validated. 
In TEF we have this scenario in ES and DE Telcos so far.

It is also understood this situation cannot (be) exists in other Telcos/countries, so we propose a way to have this feature as an optional one, so it does not compromise/impose additional logic to a Telco/Country that does not use/require it.

**Possible evolution**
Proposal is the following:

Some Basis for the proposal:
1) Not needed in 1-STEP.
-  In this way, we can also remove validationURL from 1-STEP.

2) Only used in 2-STEP Payment when the validation information is provided. So optional due to Business Requirements
**In preparePayment response**, define new param “validationInfo” (name open, maybe be also for instance otpValidation or paymentValidation), with two variants:
- action: open
> validationURL -> indicates a public URL where the OTP can be input and validated. This is a Backend generated URL so no customizable by the merchant

image

- action: validate
> authorizationId -> the identifier for this payment authorization -> This is the case where merchant builds the UI related to the paymentValidation so an endpoint is needed for the merchant to invoke the validation 

image

**NEW validatePayment** endpoint which would have as inputs:
o	code: OTP received out-of-band
o	authorizationId, as provided by API en preparePayment

Then validatePayment would look something like on this style:

Request 
image

Response
image

Additional Points:
**confirmPayment** we would need and extra Exception for the case the payment is not validated yet
- paymentId is not validated yet ("code": "CARRIER_BILLING.PAYMENT_NOT_VALIDATED","message": "Payment has not been validated").
 
- Probably we may need an extra state "waiting for confirmation" (name in API TBD) to cover with this optional process
Some initial input

•	If preparePayment is SYNC and no validationInfo is present -> state is “reserved”
•	If preparePayment is SYNC and validationInfo is present -> state is “waiting for validation” -> when validated it transitions to “reserved”
•	If preparePayment is ASYNC and no validationInfo is present -> status is “processing” -> when process ends it transitions to “reserved”
•	If preparePayment is ASYNC and validationInfo is present -> state is “waiting for validation” -> when validated it transitions to “reserved”

**Additional context**
N/A

### LINKED ISSUE #117: Business Exception for a given Payment Procedure
**Problem description**
Busines Scenario where a payment cannot be allowed either via 1-STEP or 2-STEP, due to business considerations with a Merchant or Hyperscaler/Aggregator or even underlying procedures within a Telco in a given country.

Therefore, it is proposed a new Exception to cover this scenario in both mechanisms (1-STEP, 2-STEP)

**Possible evolution**
Add new exception 403 "CARRIER_BILLING.PAYMENT_NOT_ALLOWED" in endpoints:
- createPayment
- preparePayment

**Alternative solution**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by PedroDiez**
#### What type of PR is this?

Add one of the following kinds:
* correction
* enhancement/feature
* documentation

#### What this PR does / why we need it:

This PR deals with some Issues whose output is already aligned/agreed:

* [Issue#109](https://github.com/camaraproject/CarrierBillingCheckOut/issues/109) [COVERED]
> Enhancements in filtering for `retrievePayments`

* [Issue#110](https://github.com/camaraproject/CarrierBillingCheckOut/issues/110) [COVERED]
> Aligments in transitionOperationStatus 
> New Exception 403 "CARRIER_BILLING.PAYMENT_DENIED"
> New event payment-denied
> Enhancements in documentation. 

* [Issue#111](https://github.com/camaraproject/CarrierBillingCheckOut/issues/111) [COVERED]
> Clarifications for `retrievePayments` and `retrievePayment` endpoints
> New Exception 403 "CARRIER_BILLING.INVALID_TOKEN_CONTEXT" for these endpoints 

* [Issue#116](https://github.com/camaraproject/CarrierBillingCheckOut/issues/116) [COVERED]
> New endpoint (optional in 2-step) `validatePayment` 
> New event payment-pending-validation
> New transationOperationStatus `pending_validation`
> Enhancements in documentation. 

* [Issue#117](https://github.com/camaraproject/CarrierBillingCheckOut/issues/117) [COVERED]
> Implicitly covered with Issue#110

* Other enhancements/corrections:
> Fix scopes according to model proposed in [Commonalities/PR#57](https://github.com/camaraproject/Commonalities/pull/57)
> Fix typo in transactionOperationStatus query_param enum value "defined"
> Examples values enhancement

#### Which issue(s) this PR fixes:

Fixes #109
Fixes #110 
Fixes #111 
Fixes #116 
Fixes #117 

Other enhancements/corrections:
* Fix scopes according to model proposed in [Commonalities/PR#57](https://github.com/camaraproject/Commonalities/pull/57)
* Fix typo in transactionOperationStatus query_param enum value "defined"
* Examples values enhancement

#### Special notes for reviewers:

N/A

#### Changelog input

N/A

#### Additional documentation 

N/A

### CODE REVIEW

**Review by bigludo7 — COMMENTED** [review]
One comment to discuss about the transition to Denied.

Except that look fine for me. Just wondering if we should craft 2 examples for POST/payments/prepare response : the one that you provided and another simpler (without validationInfo & webhook).

File: code/API_definitions/carrier_billing.yaml [hunk 1]
```diff
@@ -56,24 +56,40 @@ info:
     **1-STEP Payment**
 
     If `createPayment` is a **SYNC** process:
-    - Response contains `paymentId` and status=`succeeded`. In case of any error scenario `paymentId` is not created.
+    - Response contains `paymentId` and status=`succeeded`.
+    - In case of any error scenario `paymentId` is not created.
```
[the remark this answers is not in the document]
Comment (PedroDiez) [reply code/API_definitions/carrier_billing.yaml:60]: Thanks for the comment Ludovic, I will review on 11th December week

[the remark this answers is not in the document]
Comment (PedroDiez) [reply code/API_definitions/carrier_billing.yaml:60]: Understood Ludovic, in 1-STEP, SYNC behaviour, "Denied" state is not possible. If payment is denied, exception is raised instead and no paymentId is generated.

Probably the transition to Denied is only valid for Async flow -> Yes, it is needed to make some adjustment in Diagram

[the remark this answers is not in the document]
Comment (PedroDiez) [reply code/API_definitions/carrier_billing.yaml:60]: Done in [PR#128](https://github.com/camaraproject/CarrierBillingCheckOut/pull/128)

**Comment by PedroDiez** [comment]
Merging this PR and I generate new Issue with related PR to set this proposal as Release Candidate

**Pull request merged**